### PR TITLE
Update AbiWinRT semantics to match CppWinRT

### LINKED ISSUE #654: Make ABI header consumption easy
With Project Reunion and other components being moved out of the OS, there needs to be a method for them to consume ABI headers for customers that have already done so.

It could be the responsibility of every component author to ship headers, but in the case of WinUI, we have dependencies on types defined in other headers, and so there isn't always a clear line on which headers we need to include in our nuget for them to build. As more components are lifted, I'd imagine that this exercise would become more confusing.

Rather than putting the onerous on each component author to generate the headers and include them, the `Microsoft.Windows.AbiWinRT` nuget package should behave the same way that `Microsoft.Windows.CppWinRT` does, and sniff out the references and generate the appropriate headers.

### PR DESCRIPTION
**Pull request by stevenbrix**
This change makes the `Microsoft.Windows.AbiWinRT` package behave similarly to `Microsoft.Windows.CppWinRT`. When a projects adds a nuget reference to `Microsoft.Windows.AbiWinRT`, the references will be sniffed out and the appropriate header files will be generated. In order to support older projects that haven't switch to modern idl, I added the ability to generate the .idl files from the referened winmd by invoking `winmdidl` before we run midl. This is only needed because of the bug described here https://github.com/microsoft/xlang/issues/654#issuecomment-647632160

Note that I removed anything from cpp/winrt that was related to `XamlMetadataProvider`, or had anything to do with `Page` or `ApplicationDefinition`, like the heap enforcement target.

## Usage

|  Property |  Description |
|-----------|---------------|
| `AbiWinRTAddAbiNamespacePrefix` | Adds the `ABI::` prefix to all generated headers. Defaults to `true`. |
| `AbiWinRTEnsureSDKHeaderCompat` | Passes the `-lowercase-include-guard` switch to `abi.exe` to enable compat with SDK headers. Defaults to `false`. |

Fixes #654

### CODE REVIEW

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 1]
```diff
@@ -6,8 +6,500 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <AbiWinRTUseHardlinksIfPossible Condition="'$(AbiWinRTUseHardlinksIfPossible)' == ''">false</AbiWinRTUseHardlinksIfPossible>
+        <AbiWinRTWriteOnlyWhenDifferent Condition="('$(AbiWinRTWriteOnlyWhenDifferent)' == '') And (('$(MSBuildToolsVersion)' == 'Current') Or ('$(MSBuildToolsVersion)' &gt;= '15'))">true</AbiWinRTWriteOnlyWhenDifferent>
+        <AbiWinRTWriteOnlyWhenDifferent Condition="'$(AbiWinRTWriteOnlyWhenDifferent)' != 'true'">false</AbiWinRTWriteOnlyWhenDifferent>
+        <AbiWinRTPath Condition="'$(AbiWinRTPath)'==''">$(MSBuildThisFileDirectory)..\..\bin\</AbiWinRTPath>
+
+        <!-- Default modern idl semantics to false considering that most users probably have not made the switch to the modern idl syntax -->
+        <AbiWinRTModernIDL Condition="'$(AbiWinRTModernIDL)'==''">false</AbiWinRTModernIDL>
+
+        <!-- Only generate the platform and component projection if using modern idl. Otherwise, allow the legacy
+             toolchain to kick in, as well as use the existing references from the Windows SDK -->
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:30]: update comments copied from c++/winrt

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 2]
```diff
@@ -6,8 +6,500 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <AbiWinRTAddAbiNamespacePrefix Condition="'$(AbiWinRTAddAbiNamespacePrefix)'==''">true</AbiWinRTAddAbiNamespacePrefix>
+        <AbiWinRTCommandAddAbiNamespacePrefix Condition="'$(AbiWinRTAddAbiNamespacePrefix)'=='true'">-ns-prefix always</AbiWinRTCommandAddAbiNamespacePrefix>
+
+        <!-- If not generating the platform headers, then default to ensure that there is compat -->
+        <AbiWinRTEnsureSDKHeaderCompat Condition="'$(AbiWinRTEnsureSDKHeaderCompat)'=='' and '$(AbiWinRTEnablePlatformProjection)'=='false'">true</AbiWinRTEnsureSDKHeaderCompat>
+        <AbiWinRTCommandEnsureSDKHeaderCompat Condition="'$(AbiWinRTEnsureSDKHeaderCompat)'=='true'">-lowercase-include-guard</AbiWinRTCommandEnsureSDKHeaderCompat>
+
+        <GeneratedFilesDir Condition="'$(GeneratedFilesDir)' == ''">$(IntDir)Generated Files\</GeneratedFilesDir>
+        <!--Override SDK's uap.props setting to ensure version-matched headers-->
+        <AbiWinRT_IncludePath>$(GeneratedFilesDir)</AbiWinRT_IncludePath>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:45]: >AbiWinRT_IncludePath [](start = 9, length = 20)

review unused properties and remove

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 3]
```diff
@@ -6,8 +6,500 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+      <Midl Condition="'$(AbiWinRTModernIDL)' != 'false'">
+        <AdditionalMetadataDirectories Condition="'%(AdditionalMetadataDirectories)' == '' And '$(WindowsSDK_MetadataFoundationPath)' != ''">$(WindowsSDK_MetadataFoundationPath);%(AdditionalMetadataDirectories)</AdditionalMetadataDirectories>
+        <AdditionalMetadataDirectories Condition="'%(AdditionalMetadataDirectories)' == '' And '$(WindowsSDK_MetadataFoundationPath)' == ''">$(WindowsSDK_MetadataPath);%(AdditionalMetadataDirectories)</AdditionalMetadataDirectories>
+        <AdditionalOptions>%(AdditionalOptions) /nomidl</AdditionalOptions>
+      </Midl>
+      <Link>
+        <AdditionalDependencies Condition="'$(AbiWinRTLibs)' != 'false'">%(AdditionalDependencies);WindowsApp.lib</AdditionalDependencies>
+      </Link>
+    </ItemDefinitionGroup>
+
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:503]: remove all this stuff that's cppwinrt carryover

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 1 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:30]: Hey... I wrote this comment myself :)

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 3 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:503]: Sure. For my own curiosity, why was this here? I wasn't 100% sure what `CppWinRTLibs` was for, is it for desktop projects that also want to link against WindowsApp.lib to access WinRT APIs?

Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:496]: I'd say if you're not sure what it is, remove it.  You probably only need ~100 lines in this file.  I don't think you need to be configuring other tools like cl, midlrt, mdmerge, etc.  And you don't need all the platform/component projection targeting.

---
In reply to: [443886013](https://github.com/microsoft/xlang/pull/659#discussion_r443886013) [](ancestors = 443886013)

File: src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          Returns="@(AbiWinRTDirectWinMDReferencesForIdl);@(AbiWinRTDirectWinMDReferenceDirs)">
+    <ItemGroup>
+      <_AbiWinRTDirectWinMDReferencesForIdl Remove="@(_AbiWinRTDirectWinMDReferencesForIdl)" />
+      <_AbiWinRTDirectWinMDReferencesForIdl Include="@(Reference)" Condition="'%(Reference.Extension)' == '.winmd' and '%(Reference.IsSystemReference)' != 'true'" />
+      <AbiWinRTDirectWinMDReferencesForIdl Remove="@(AbiWinRTDirectWinMDReferencesForIdl)"/>
+      <AbiWinRTDirectWinMDReferencesForIdl Include="@(_AbiWinRTDirectWinMDReferencesForIdl)">
+        <WinMDPath>%(FullPath)</WinMDPath>
+      </AbiWinRTDirectWinMDReferencesForIdl>
+
+      <!-- Generate list of reference directories to pass into winmdidl.exe -->
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets:24]: tool name is abi.exe

File: src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+  <!--Build reference IDL from WinMD project references and dynamic library project references
+      Note that Condition is evaluated before DependsOnTargets are run -->
+  <Target Name="AbiWinRTMakeReferenceIdl"
+          Condition="'@(AbiWinRTDirectWinMDReferencesForIdl)' != ''"
+          AfterTargets="GetAbiWinRTDirectWinMDReferencesForIdl"
+          Inputs="@(AbiWinRTDirectWinMDReferencesForIdl)"
+          Outputs="@(_ExpectedGeneratedIdl)">
+    <PropertyGroup>
+      <WindowsSdkToolLocation Condition="'$(WindowsSdkToolLocation)'==''">$(WDKBinRoot)\$(PreferredToolArchitecture)</WindowsSdkToolLocation>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets:57]: this is for OS builds - you can remove it and all references to it

File: src/package/abi/Microsoft.Windows.AbiWinRT.props [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    <PropertyGroup>
+        <!-- Only do this for MSBuild versions below 16.0
+             as it is since done automatically, see https://github.com/microsoft/msbuild/pull/3605-->
+        <MSBuildAllProjects Condition="'$(MSBuildToolsVersion)'  &lt;= '15'">$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+        
+        <UsingAbiWinRT>true</UsingAbiWinRT>
+    </PropertyGroup>
+
+     <PropertyGroup>
+        <PreferredToolArchitecture>x64</PreferredToolArchitecture>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.props:17]: this is cppwinrt-specific

File: src/package/abi/Microsoft.Windows.AbiWinRT.props [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        <CanReferenceWinRT>true</CanReferenceWinRT>
+        <IsNativeLanguage>true</IsNativeLanguage>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.props:19]: why is this needed?

File: src/package/abi/Microsoft.Windows.AbiWinRT.props [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        <!-- This causes VS to add the platform WinMD references for the target SDK -->
+        <WinMDAssembly>true</WinMDAssembly>
+        <!--Set a value to prevent SDK's uap.props from setting it and pointing to the wrong headers-->
+        <AbiWinRT_IncludePath>PreventSdkUapPropsAssignment</AbiWinRT_IncludePath>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.props:23]: make sure you test consuming projects to ensure they don't have colliding SDK ABI headers - that's why this was added for cppwinrt

File: src/package/abi/Microsoft.Windows.AbiWinRT.props [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    </PropertyGroup>
+
+    <ItemDefinitionGroup>
+        <ClCompile>
+            <CompileAsWinRT Condition="'%(ClCompile.CompileAsWinRT)' == ''">false</CompileAsWinRT>
+        </ClCompile>
+        <Midl Condition="'$(AbiWinRTModernIDL)' != 'false'">
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.props:30]: can you remove AbiWinRTModernIDL and only add it if necessary later (cppwinrt needed it) - just assume modern is all that's supported?

Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.props:30]: the scenarios for the abi nuget are likely to be much narrower than for cppwinrt, so the need for this degree of flexibility is much lower.

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 10]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+             as it is since done automatically, see https://github.com/microsoft/msbuild/pull/3605-->
+        <MSBuildAllProjects Condition="'$(MSBuildToolsVersion)'  &lt;= '15'">$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
     </PropertyGroup>
 
-</Project>
+    <PropertyGroup>
+        <AbiWinRTVerbosity Condition="'$(AbiWinRTVerbosity)' == ''">normal</AbiWinRTVerbosity>
+        <AbiWinRTCommandVerbosity Condition="'$(AbiWinRTVerbosity)' == 'high'">-verbose</AbiWinRTCommandVerbosity>
+        <AbiWinRTProjectWinMD>$(OutDir)$(RootNamespace).winmd</AbiWinRTProjectWinMD>
+        <AbiWinRTMergedDir>$(IntDir)Merged\</AbiWinRTMergedDir>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:18]: how does all this configuration interact with the cppwinrt nuget?  seems like they would collide badly

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 11]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <AbiWinRTUnmergedDir>$(IntDir)Unmerged\</AbiWinRTUnmergedDir>
+        <AbiWinRTSkipUnchangedFiles Condition="'$(AbiWinRTSkipUnchangedFiles)' == ''">true</AbiWinRTSkipUnchangedFiles>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:20]: skip unchanged files and hardlinks were made optional for cppwinrt to support older versions of msbuild.  I'd recommend removing all this flexibility and asserting your minimal version requirement

File: src/package/abi/Microsoft.Windows.AbiWinRT.props [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            <EnableWindowsRuntime>true</EnableWindowsRuntime>
+            <MetadataFileName>$(IntDir)Unmerged\%(Filename).winmd</MetadataFileName>
+            <GenerateClientFiles Condition="'%(Midl.GenerateClientFiles)'==''">None</GenerateClientFiles>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.props:33]: cppwinrt doesn't support proxy/stub generation, which is why these are all disabled.  I don't think you want to do that for abi.exe

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 13]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <AbiWinRTPath Condition="'$(AbiWinRTPath)'==''">$(MSBuildThisFileDirectory)..\..\bin\</AbiWinRTPath>
+
+        <!-- Default modern idl semantics to false considering that most users probably have not made the switch to the modern idl syntax -->
+        <AbiWinRTModernIDL Condition="'$(AbiWinRTModernIDL)'==''">false</AbiWinRTModernIDL>
+
+        <!-- Only generate the platform and component projection if using modern idl. Otherwise, allow the legacy
+             toolchain to kick in, as well as use the existing references from the Windows SDK -->
+        <AbiWinRTEnableComponentProjection Condition="'$(AbiWinRTEnableComponentProjection)' == '' and '$(AbiWinRTModernIdl)'!='false'">true</AbiWinRTEnableComponentProjection>
+        <AbiWinRTEnablePlatformProjection Condition="'$(AbiWinRTEnablePlatformProjection)' == '' and '$(AbiWinRTModernIdl)'!='false'">true</AbiWinRTEnablePlatformProjection>
+        <AbiWinRTEnableReferenceProjection Condition="'$(AbiWinRTEnableReferenceProjection)' == '' and '$(AbiWinRTModernIdl)'!='false'">true</AbiWinRTEnableReferenceProjection>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:31]: cppwinrt makes a distinction between platform, component, and reference projections which abi.exe should not do.  there should only be a single combined platform/reference projection here (component is a unique cppwinrt concept implying authoring support, not just consuming a third party lib).  and there should be a build assertion to prevent contamination with SDK abi headers.  maybe a static assert or pragma detect mismatch would help

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 14]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+
+        <!-- Default ABI headers to always be prepended with the ABI namespace -->
+        <AbiWinRTAddAbiNamespacePrefix Condition="'$(AbiWinRTAddAbiNamespacePrefix)'==''">true</AbiWinRTAddAbiNamespacePrefix>
+        <AbiWinRTCommandAddAbiNamespacePrefix Condition="'$(AbiWinRTAddAbiNamespacePrefix)'=='true'">-ns-prefix always</AbiWinRTCommandAddAbiNamespacePrefix>
+
+        <!-- If not generating the platform headers, then default to ensure that there is compat -->
+        <AbiWinRTEnsureSDKHeaderCompat Condition="'$(AbiWinRTEnsureSDKHeaderCompat)'=='' and '$(AbiWinRTEnablePlatformProjection)'=='false'">true</AbiWinRTEnsureSDKHeaderCompat>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:38]: maybe this allows combining with sdk headers?  if so, then the suggested build assert could be conditional

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 15]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <AbiWinRTCommandEnsureSDKHeaderCompat Condition="'$(AbiWinRTEnsureSDKHeaderCompat)'=='true'">-lowercase-include-guard</AbiWinRTCommandEnsureSDKHeaderCompat>
+
+        <GeneratedFilesDir Condition="'$(GeneratedFilesDir)' == ''">$(IntDir)Generated Files\</GeneratedFilesDir>
+        <!--TEMP: Override NuGet SDK's erroneous setting in uap.props -->
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:42]: this may no longer be needed - test with nuget SDK distro (and/or ask @manodasanW )

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 16]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <!-- Note: Before* targets run before Compute* targets. -->
+        <ComputeMidlInputsTargets>
+            $(ComputeMidlInputsTargets);AbiWinRTSetMidlReferences;
+        </ComputeMidlInputsTargets>
+        <AfterMidlTargets>
+            $(AfterMidlTargets);
+            GetAbiWinRTMdMergeInputs;
+            AbiWinRTMergeProjectWinMDInputs;
+            GetResolvedWinMD;
+            AbiWinRTCopyWinMDToOutputDirectory;
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:60]: this all feels to 'intrusive' - shouldn't this package just perform a 'read-only' scan of a project's winmds (direct, referenced, etc) and create headers?

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 17]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        </AfterMidlTargets>
+        <ResolveAssemblyReferencesDependsOn>
+            $(ResolveAssemblyReferencesDependsOn);GetAbiWinRTProjectWinMDReferences;AbiWinRTMarkStaticLibrariesPrivate;
+        </ResolveAssemblyReferencesDependsOn>
+        <!-- Note: Before* targets run before Compute* targets. -->
+        <BeforeClCompileTargets>
+            $(BeforeClCompileTargets);AbiWinRTMakeProjections;
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:67]: all ClCompile stuff should be removed - cppwinrt specific

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 18]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        </BeforeClCompileTargets>
+        <CleanDependsOn>
+            $(CleanDependsOn);AbiWinRTClean
+        </CleanDependsOn>
+
+    </PropertyGroup>
+
+    <!-- For a static library we don't want the winmd/lib/pdb to be packaged -->
+    <PropertyGroup Condition="'$(ConfigurationType)' == 'StaticLibrary'">
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:76]: remove

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 19]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <IncludeCopyWinMDArtifactsOutputGroup>false</IncludeCopyWinMDArtifactsOutputGroup>
+    </PropertyGroup>
+
+    <Target Name="AbiWinRTVerifyKitVersion" Condition="'$(AbiWinRTOverrideSDKReferences)' != 'true'">
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:80]: remove all references to AbiWinRTOverrideSDKReferences (cppwinrt carryover for os build support)

Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:80]: I'd remove this target - can always add msbuild assertions later if needed

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 20]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+            <_FilesToDelete Include="$(AbiWinRTUnmergedDir)**"/>
+        </ItemGroup>
+        <Delete Files="@(_FilesToDelete)"/>
+    </Target>
+
+    <!-- Target used only to evaluate AbiWinRTGenerateWindowsMetadata if it doesn't already have a value -->
+    <Target Name="AbiWinRTComputeGenerateWindowsMetadata"
+            DependsOnTargets="GetAbiWinRTProjectWinMDReferences;$(AbiWinRTComputeGenerateWindowsMetadataDependsOn)">
+
+        <PropertyGroup>
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:106]: this nuget package is "consume only" - referenced by clients of winmds - all static lib concerns should go

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 21]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+                <Implementation>$(WinMDImplementationPath)$(TargetName)$(TargetExt)</Implementation>
+                <FileType>winmd</FileType>
+                <WinMDFile>true</WinMDFile>
+                <ProjectType>$(ConfigurationType)</ProjectType>
+            </WinMDFullPath>
+
+            <Link Remove="@(Link)" Condition="'%(Link.DeleteSoon)' == 'true'" />
+        </ItemGroup>
+
+        <!-- Add C++/WinRT primary WinMD to the WinMDFullPath if AbiWinRTGenerateWindowsMetadata is true -->
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:160]: update all c++/winrt comments

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 22]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+
+        <Message Text="GetResolvedWinMD: @(WinMDFullPath->'%(FullPath)')" Importance="$(AbiWinRTVerbosity)"/>
+    </Target>
+
+    <!-- Static library reference WinMDs are merged into the project WinMD that
+         references it and might have the same name because they often share namespace. 
+         Therefore they shouldn't be copied to the output folder
+         because they might override files in the output folder with the
+         same name, causing missing types. -->
+    <Target Name="AbiWinRTMarkStaticLibrariesPrivate"
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:181]: remove

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 23]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        </ItemGroup>
+        <Message Text="AbiWinRTDirectWinMDReferences: @(AbiWinRTDirectWinMDReferences->'%(WinMDPath)')" Importance="$(AbiWinRTVerbosity)"/>
+    </Target>
+
+    <!--Get direct WinMD project references for projections, IDL processing, and AppX packaging-->
+    <Target Name="GetAbiWinRTProjectWinMDReferences"
+            DependsOnTargets="ResolveProjectReferences;$(GetAbiWinRTProjectWinMDReferencesDependsOn)"
+            Returns="@(AbiWinRTStaticProjectWinMDReferences);@(AbiWinRTDynamicProjectWinMDReferences)">
+        <ItemGroup>
+            <!-- Get static library project references -->
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:244]: remove

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 24]
```diff
@@ -6,8 +6,493 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+    <Target Name="GetAbiWinRTMdMergeInputs"
+                DependsOnTargets="AbiWinRTResolveReferences;"
+                Returns="@(AbiWinRTMdMergeMetadataDirectories);@(AbiWinRTMdMergeInputs)">
+        <ItemGroup>
+            <_MdMergeInputs Remove="@(_MdMergeInputs)"/>
+            <_MdMergeInputs Include="@(Midl)">
+                <WinMDPath>%(Midl.OutputDirectory)%(Midl.MetadataFileName)</WinMDPath>
+                <MdMergeOutputFile>$(AbiWinRTProjectWinMD)</MdMergeOutputFile>
+            </_MdMergeInputs>
+            <!-- Static libraries don't mdmerge other static libraries.
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:282]: remove

File: src/package/abi/Microsoft.Windows.AbiWinRT.props (hunk 9 above)
Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.props:30]: now that I think of it, if this nuget is for consumption only, then all midl/mdmerge configuration should go away (like clcompile/link) - you should only be scanning for referenced winmds to generated headers

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 17 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:67]: Some amount of `ClCompile` stuff is needed, this is for generating the headers. You also need to update the include paths as well.

File: src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets (hunk 5 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets:57]: which one/?

File: src/package/abi/Microsoft.Windows.AbiWinRT.props (hunk 8 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.props:23]: I've removed this since it isn't needed then. I don't see any collision, the files generated by abiwinrt are consumed

File: src/package/abi/Microsoft.Windows.AbiWinRT.props (hunk 7 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.props:19]: it's not, i've removed it

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 14 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:38]: which build assert?

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 15 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:42]: Doesn't look like it is

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 16 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:60]: yup, i've updated it since we talked and removed this

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 10 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:18]: this nuget provides an error when it detects cpp/winrt nuget package is also being used

File: src/package/abi/Microsoft.Windows.AbiWinRT.props (hunk 7 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.props:19]: it's not

File: src/package/abi/Microsoft.Windows.AbiWinRT.props (hunk 9 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.props:30]: going to just resolve all these midl related ones since we agreed a different nuget package is appropriate for that

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 13 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:31]: Does the component scenario not make sense for abi.exe since midlrt generates the same headers?

Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:31]: how do you prevent contamination with sdk abi headers? we could probably remove the sdk abi header location from `ClCompile.AdditionalIncludeDirectories`?

File: src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets (hunk 5 above)
Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets:57]: n.m - misread this

---
In reply to: [453766996](https://github.com/microsoft/xlang/pull/659#discussion_r453766996) [](ancestors = 453766996)

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 13 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:31]: actually I don't think we can do that, since wrl headers are also in that location

File: src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets (hunk 4 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.IdlGen.targets:24]: this was for running winmidl, but is no longer part of this change

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 13 above)
Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets]: talk to @manodasanW - he had to tweak the SDK Nuget to prevent collisions with the uap.props additional include dirs

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 14 above)
Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:24]: can't remember where that comment was - but somewhere you could assert (cause a build warning/error) that ensures SDK WRL headers don't accidentally mix with ABIWinRT headers

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 17 above)
Comment (Scottj1s) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:41]: Yes, additional include path is fine (but nothing else - no /await, /std:c++17, etc)

Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets:41]: gotcha, sweet - yeah i've removed all of that

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets (hunk 13 above)
Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets]: The way I did it was by pre-pending `GeneratedFilesDir` to the `ClCompile.AdditionalIncludeDirectories`. I'll follow up with Mano to see what he suggests. I'd love for an actual assert that isn't dependent on MSBuild import order.

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 25]
```diff
@@ -6,8 +6,155 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+        <BeforeClCompileTargets>
+            $(BeforeClCompileTargets);AbiWinRTMakeProjections;
+        </BeforeClCompileTargets>
+        <CleanDependsOn>
+            $(CleanDependsOn);AbiWinRTClean
+        </CleanDependsOn>
+
+    </PropertyGroup>
+
+    <!-- Using both cpp/winrt and abi/winrt together is unsupported. -->
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:49]: C++/WinRT (or cppwinrt, but not both)

File: src/tool/abi/readme.md [hunk 26]
```diff
@@ -0,0 +1,16 @@
+## The ABI/WinRT language projection
+
+The Application Binary Interface (ABI) projection generates header files for consuming WinRT APIs that are more familiar to headers generated from COM .idl files.
+
+**Note: This tool is for legacy codebases and not recommended for new development. C++ developers should instead use https://github.com/microsoft/cppwinrt, which supports standard c++ and the latest c++ language features.**
```
Comment (Scottj1s) [inline src/tool/abi/readme.md:5]: capitalize C++

File: src/tool/abi/readme.md [hunk 27]
```diff
@@ -0,0 +1,16 @@
+## The ABI/WinRT language projection
+
+The Application Binary Interface (ABI) projection generates header files for consuming WinRT APIs that are more familiar to headers generated from COM .idl files.
```
Comment (Scottj1s) [inline src/tool/abi/readme.md:3]: "more familiar" -> "similar"

File: src/package/abi/Microsoft.Windows.AbiWinRT.targets [hunk 28]
```diff
@@ -46,7 +46,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
     </PropertyGroup>
 
-    <!-- Using both cpp/winrt and abi/winrt together is unsupported. -->
+    <!-- Using both c++/winrt and abi/winrt together is unsupported. -->
     <Target Name="AbiWinRTValidateNoCppWinRT">
         <Error Condition="'$(CppWinRTPackage)' != ''" Text="Abi/WinRT and Cpp/WinRT create competing projections and are not compatible.
```
Comment (Scottj1s) [inline src/package/abi/Microsoft.Windows.AbiWinRT.targets:51]: and this

Comment (stevenbrix) [reply src/package/abi/Microsoft.Windows.AbiWinRT.targets]: 🎵 Scott's got, the eye of the tiger... 🎵
